synapse/third_party_rules/access_rules.py

```diff
@@ -446,8 +449,130 @@ def _on_membership_or_invite(
+            return
+
+        user_id = event.get("sender")
+        if not user_id:
+            return
+
+        # Get every admin user defined in the room's state
+        admin_users = [
+            user
+            for user, power_level in power_level_content["users"].items()
+            if power_level >= 100
+        ]
+
+        if user_id not in admin_users:
+            # This user is not an admin, ignore them
+            return
+
+        # Filter these users to only those who are actually joined or invited to the room
+        joined_members = [
+            user_id
+            for (event_type, user_id), event in state_events.items()
+            if event_type == EventTypes.Member
+            and event.membership in [Membership.JOIN, Membership.INVITE]
+        ]
+        admin_users = [user for user in admin_users if user in joined_members]
+
+        if len(admin_users) > 1:
+            # There's another admin user in, or invited to, the room
+            return
+
+        # Modify the existing power levels to raise all required types to 100
+        #
+        # This changes a power level state event's content from something like:
+        # {
+        #     "redact": 50,
+        #     "state_default": 50,
+        #     "ban": 50,
+        #     "notifications": {
+        #         "room": 50
+        #     },
+        #     "events": {
+        #         "m.room.avatar": 50,
+        #         "m.room.encryption": 50,
+        #         "m.room.canonical_alias": 50,
+        #         "m.room.name": 50,
+        #         "im.vector.modular.widgets": 50,
+        #         "m.room.topic": 50,
+        #         "m.room.tombstone": 50,
+        #         "m.room.history_visibility": 100,
+        #         "m.room.power_levels": 100
+        #     },
+        #     "users_default": 0,
+        #     "events_default": 0,
+        #     "users": {
+        #         "@admin:example.com": 100,
+        #     },
+        #     "kick": 50,
+        #     "invite": 0
+        # }
+        #
+        # to
+        #
+        # {
+        #     "redact": 100,
+        #     "state_default": 100,
+        #     "ban": 100,
+        #     "notifications": {
+        #         "room": 50
+        #     },
+        #     "events": {}
+        #     "users_default": 0,
+        #     "events_default": 100,
+        #     "users": {
+        #         "@admin:example.com": 100,
+        #     },
+        #     "kick": 100,
+        #     "invite": 100
+        # }
+        for key, value in power_level_content.items():
+            # Do not change "users_default", as that key specifies the default power
+            # level of new users
+            if isinstance(value, int) and key != "users_default":
+                power_level_content[key] = 100
+        power_level_content["events"] = {}
+
+        # Freeze the room by raising the required power level to send events to 100
+        logger.info("Freezing room '%s'", event.room_id)
```

Comment:
these probably belong further up, at line 507ish?

Reply:
Fair yeah. I originally moved it down as it was firing even if the conditions were not met for "the last admin leaving", but looks like I overshot it.